### PR TITLE
Qt: Fix for bitcoin-qt becoming unresponsive during shutdown (issue #13217)

### PR DESCRIPTION
**Pull request by LeandroRocha84**
This pull request contains a fix for the issue #13217, where bitcoin-qt becomes unresponsive during shutdown after a long running call (like gettxoutsetinfo) in the QT console.
The problem occurs because, during shutdown, waiting for the RPCExecutor thread to finish happens in the main thread.

### CODE REVIEW

File: src/qt/rpcconsole.cpp [hunk 1]
```diff
@@ -968,19 +1014,21 @@ void RPCConsole::browseHistory(int offset)
...
     // Requests from this object must go to executor
-    connect(this, SIGNAL(cmdRequest(QString, QString)), executor, SLOT(request(QString, QString)));
+    connect(this, SIGNAL(cmdRequest(QString, QString)), executorWrapper, SLOT(request(QString, QString)));
+    connect(executorWrapper, SIGNAL(shutdownComplete()), this, SIGNAL(rpcExecutorShutdownComplete()));
 
     // On stopExecutor signal
     // - quit the Qt event loop in the execution thread
-    connect(this, SIGNAL(stopExecutor()), &thread, SLOT(quit()));
+    connect(this, SIGNAL(stopExecutor()), executorWrapper, SLOT(shutdown()));
+    //connect(this, SIGNAL(stopExecutor()), &thread, SLOT(quit()));
```
Comment (promag) [inline src/qt/rpcconsole.cpp]: Remove commented code?

**Review by promag — COMMENTED** [review]
Indeed looks like unnecessary complexity. IIUC the main even loop blocks when joining the executor thread, so consider a different approach instead?
```cpp
QEventLoop loop;
connect(&thread, SIGNAL(finished()), &loop, SLOT(quit()));
if (!thread.isFinished()) loop.exec();
```

**Comment by LeandroRocha84** [comment]
Thanks for the suggestion @promag 
I've updated the PR to remove the unnecessary complexity. Aside from your suggested code, I've made a small change to bitcoingui.cpp so that the UI clientModel is not cleared until the RPCExecutor thread joins. Otherwise, network events, like a new connection, that trigger a UI update would raise exceptions.

File: src/qt/rpcconsole.cpp [hunk 2]
```diff
@@ -688,8 +689,10 @@ void RPCConsole::setClientModel(ClientModel *model)
     if (!model) {
         // Client model is being set to 0, this means shutdown() is about to be called.
         // Make sure we clean up the executor thread
+        QEventLoop loop;
+        connect(&thread, SIGNAL(finished()), &loop, SLOT(quit()));
```
Comment (promag) [inline src/qt/rpcconsole.cpp]: Sorry, could use new connect syntax?
```cpp
connect(&thread, &QThread::finished, &loop, &QEventLoop::quit)

Comment (LeandroRocha84) [reply src/qt/rpcconsole.cpp]: Sure thing. That's done.
And rebased.

File: src/qt/bitcoingui.cpp [hunk 3]
```diff
@@ -470,13 +470,13 @@ void BitcoinGUI::setClientModel(ClientModel *_clientModel)
         }
 #endif // ENABLE_WALLET
         unitDisplayControl->setOptionsModel(_clientModel->getOptionsModel());
-        
+
```
Comment (promag) [inline src/qt/bitcoingui.cpp]: Rebase to remove this change.

File: src/qt/bitcoingui.cpp [hunk 4]
```diff
@@ -437,9 +437,9 @@ void BitcoinGUI::createToolBars()
 
 void BitcoinGUI::setClientModel(ClientModel *_clientModel)
 {
-    this->clientModel = _clientModel;
     if(_clientModel)
     {
+        this->clientModel = _clientModel;
```
Comment (promag) [inline src/qt/bitcoingui.cpp]: Why this change? Should update `this->clientModel` event if `_clientModel` is null.

File: src/qt/bitcoingui.cpp [hunk 5]
```diff
@@ -470,13 +470,13 @@ void BitcoinGUI::setClientModel(ClientModel *_clientModel)
...
         unitDisplayControl->setOptionsModel(_clientModel->getOptionsModel());
-        
+
         OptionsModel* optionsModel = _clientModel->getOptionsModel();
         if(optionsModel)
         {
             // be aware of the tray icon disable state change reported by the OptionsModel object.
             connect(optionsModel,SIGNAL(hideTrayIconChanged(bool)),this,SLOT(setTrayIconVisible(bool)));
-        
+
```
Comment (promag) [inline src/qt/bitcoingui.cpp]: Same as above.

File: src/qt/bitcoingui.cpp (hunk 4 above)
Comment (LeandroRocha84) [reply src/qt/bitcoingui.cpp]: Because now the main thread is not blocked, network events can trigger a GUI update. If the clientModel is set to 0 or null at the time, the application will crash.
Example: When number of connections change, BitcoinGUI::updateNetworkState() gets invoked and this method assumes that clientModel is valid.
With this change  BitcoinGUI clientModel is cleared only after the RPCExecutor thread joins and shutdown resumes in the main thread.

File: src/qt/bitcoingui.cpp (hunk 3 above)
Comment (LeandroRocha84) [reply src/qt/bitcoingui.cpp]: Done

File: src/qt/bitcoingui.cpp (hunk 5 above)
Comment (LeandroRocha84) [reply src/qt/bitcoingui.cpp]: Done!

**Comment by LeandroRocha84** [comment]
Hi all, please let me know if there's anything that needs to be done to have this fix merged.
Thanks

File: src/qt/rpcconsole.cpp [hunk 6]
```diff
@@ -688,8 +689,10 @@ void RPCConsole::setClientModel(ClientModel *model)
     if (!model) {
         // Client model is being set to 0, this means shutdown() is about to be called.
         // Make sure we clean up the executor thread
+        QEventLoop loop;
+        connect(&thread, &QThread::finished, &loop, &QEventLoop::quit);
         Q_EMIT stopExecutor();
-        thread.wait();
+        if (!thread.isFinished()) loop.exec();
```
Comment (hebasto) [inline src/qt/rpcconsole.cpp]: You're implicitly using `QueuedConnection` here: https://github.com/bitcoin/bitcoin/blob/9255963890f07f93646cc32703da935dcb24d454/src/qt/rpcconsole.cpp#L693
Therefore,  the `thread.isFinished()` is called immediately after the `stopExecutor()` signal is emitted. If the `thread` is idle the value of `thread.isFinished()` may be unpredictable, IMO.

Comment (LeandroRocha84) [reply src/qt/rpcconsole.cpp]: If I understood your concern correctly, you're afraid that the thread.isFinished() result may be unpredictable and as a result, the QEventLoop may never quit since it is tied to the thread finished() signal.

I don't see a problem considering how the QThread lifecycle works. As per documentation, the order is:
* Thread quit() or exit() is called
* As soon as exec completes, isFinished starts to return true
* The finished() signal is emitted

As a result we should consistently get the loop.exec before the thread finished signal is emitted.
Also, if thread.isFinished return true, the code in the if statement is not executed and shutdown continues normally.

Please let me know if I misunderstood your concern.
Thanks,

Comment (LeandroRocha84) [reply src/qt/rpcconsole.cpp]: As you noticed, we're using `QueuedConnection` implicitly here.
The EventLoop will only receive the thread finished event after we call `exec()`. I ran multiple tests locally to confirm that. Even when the thread finishes right away and we insert some long running process before the exec, it still exits successfully.
That would in fact be a problem if we were using Qt::DirectConnection.

@promag , do you have anything to add to this comment?  (as you suggested that snippet of code)

Comment (hebasto) [reply src/qt/rpcconsole.cpp]: Agree: forcing event dispatching by means of a local event loop works as expected.

File: src/qt/bitcoingui.cpp [hunk 7]
```diff
@@ -488,8 +488,11 @@ void BitcoinGUI::setClientModel(ClientModel *_clientModel)
             // Disable context menu on tray icon
             trayIconMenu->clear();
         }
-        // Propagate cleared model to child objects
+        // Propagate cleared model to RPC console
         rpcConsole->setClientModel(nullptr);
+
+        // Once RPC console is cleared, propagate to other child objects
+        this->clientModel = 0;
```
Comment (hebasto) [inline src/qt/bitcoingui.cpp]:
```
this->clientModel = nullptr;
```

**Review by hebasto — CHANGES_REQUESTED** [review]

File: src/qt/bitcoingui.cpp (hunk 7 above)
Comment (LeandroRocha84) [reply src/qt/bitcoingui.cpp]: Done. Changed to nullptr.
I had it set to 0 before because it is the current behavior in master. `BitcoinApplication::requestShutdown()` in bitcoin.cpp invokes `window->setClientModel(0)`

**Comment by LeandroRocha84** [comment]
All comments addressed. Commits squashed. Please let me know if there's anything.

**Comment by hebasto** [comment]
re-tACK abff36969da9998e36c72f2df7faf68564f0c8db (Linux, macOS 10.13.6)

**Comment by jonasschnelli** [comment]
@LeandroRocha84 sorry for keeping this up so long...
can you rebase once and then we try to get this merged?

**Comment by LeandroRocha84** [comment]
Thanks @jonasschnelli I'm working on solving the conflicts after the rebase.
The order change in fd6d499 breaks the solution in this PR as it makes the QEventLoop.exec returns right away. I've confirmed that the bug this PR intends to solve is still reproducible; the shutdown screen is frozen/blocked when we have a long running RPC command.
I'm going to spend some time this weekend to find out how exactly the order change I linked affects the QEventLoop.

Adding @promag as he might have some insight.

**Comment by LeandroRocha84** [comment]
Conflicts resolved, rebase completed. I've tested the fix on linux.
@jonasschnelli please let me know if there's anything else I can do to get this pull request completed.
Thank you.

File: src/qt/bitcoingui.cpp (hunk 4 above)
Comment (promag) [reply src/qt/bitcoingui.cpp]: > BitcoinGUI::updateNetworkState() gets invoked and this method assumes that clientModel is valid.

Then those need to be fixed IMO. With your change calling `setClientModel(nullptr)` on the `BitcoinGUI` instance is a noop.

File: src/qt/rpcconsole.cpp [hunk 8]
```diff
@@ -686,8 +687,11 @@ void RPCConsole::setClientModel(ClientModel *model)
     }
     if (!model) {
         // Client model is being set to 0, this means shutdown() is about to be called.
+        QEventLoop loop;
+        connect(&thread, &QThread::finished, &loop, &QEventLoop::quit);
         thread.quit();
-        thread.wait();
+        if (!thread.isFinished())
```
Comment (promag) [inline src/qt/rpcconsole.cpp]: Add braces `{`.

File: src/qt/bitcoingui.cpp [hunk 9]
```diff
@@ -1311,7 +1314,8 @@ void BitcoinGUI::detectShutdown()
     {
         if(rpcConsole)
             rpcConsole->hide();
-        qApp->quit();
+        if(!closeEventRegistered)
```
Comment (promag) [inline src/qt/bitcoingui.cpp]: Should be
```cpp
if (!closeEventRegistered) qApp->quit();
```
Or
```cpp
if (!closeEventRegistered) {
    qApp->quit();
}
```

File: src/qt/bitcoingui.h [hunk 10]
```diff
@@ -161,6 +161,7 @@ class BitcoinGUI : public QMainWindow
     RPCConsole* rpcConsole = nullptr;
     HelpMessageDialog* helpMessageDialog = nullptr;
     ModalOverlay* modalOverlay = nullptr;
+    bool closeEventRegistered = false;
```
Comment (promag) [inline src/qt/bitcoingui.h]: This is used to prevent calling `qApp->quit()` more than once. Is this a real problem - could this be removed?

**Review by promag — CHANGES_REQUESTED** [review]

File: src/qt/bitcoingui.cpp (hunk 9 above)
Comment (LeandroRocha84) [reply src/qt/bitcoingui.cpp]: Done.

File: src/qt/bitcoingui.h (hunk 10 above)
Comment (LeandroRocha84) [reply src/qt/bitcoingui.h]: That's right. This is needed since the reordering in fd6d499. The second call to ```aApp-quit()``` would make the QEventLoop exit before the RPC executor thread finishes causing the application to crash.
I've made sure that closing bitcoin QT by typing in 'stop' in the console still works as intended when this line was introduced in 9a2305a

File: src/qt/rpcconsole.cpp [hunk 11]
```diff
@@ -686,8 +687,12 @@ void RPCConsole::setClientModel(ClientModel *model)
     }
     if (!model) {
         // Client model is being set to 0, this means shutdown() is about to be called.
+        QEventLoop loop;
+        connect(&thread, &QThread::finished, &loop, &QEventLoop::quit);
         thread.quit();
-        thread.wait();
+        if (!thread.isFinished()) {
```
Comment (promag) [inline src/qt/rpcconsole.cpp]: Between `thread.isFinished()` and `loop.exec()` the thread can terminate and then `loop` won't quit.

Comment (LeandroRocha84) [reply src/qt/rpcconsole.cpp]: As we're implicitly using QueuedConnection, the thread finished signal won't be produced until after ```loop.exec()``` is called, that's my understanding based on the docs. Please let me know if you think I'm wrong here - I've tested this a lot, including with varying delays between ```thread.isFinished()``` and ```loop.exec()```
Another comment I wrote on this PR in a thread with @hebasto : https://github.com/bitcoin/bitcoin/pull/13674#discussion_r231021109

Let me know whether you still think this is a problem.

Comment (promag) [reply src/qt/rpcconsole.cpp]: You are correct, I've verified `QThread::finished` signal is emitted from a different thread which defers the `QEventLoop::quit` execution.

However, considering #15313, I'm not sure if nested event loops should be avoided. Sorry for suggesting this approach but #15313 raised some concerns in this regard.

An alternative is to expose a signal `RPCConsole::finished` to `BitcoinGUI`.

Comment (LeandroRocha84) [reply src/qt/rpcconsole.cpp]: I see the point. That's done, I've exposed the thread finished signal to ````BitcoinGUI````
Can you please review?
Thanks,
